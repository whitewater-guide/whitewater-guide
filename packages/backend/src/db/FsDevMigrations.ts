// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { FsMigrations } from 'knex/lib/migrate/sources/fs-migrations';

// When restoring production dump on development, migration names will have '.js' extension
// Therefore, development migrations with '.ts' will conflict them
// this is the workaround
// https://github.com/knex/knex/blob/master/lib/migrate/sources/fs-migrations.js
export class FsDevMigrations extends FsMigrations {
  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
  constructor(
    migrationDirectories: string | string[],
    sortDirsSeparately?: boolean,
    loadExtensions?: string[],
  ) {
    super(migrationDirectories, sortDirsSeparately, loadExtensions);
  }

  getMigrationName(migration: any): string {
    return migration.file.replace('.ts', '.js');
  }
}
