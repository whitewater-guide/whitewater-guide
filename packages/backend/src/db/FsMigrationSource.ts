import fs from 'node:fs/promises';
import path from 'node:path';

import type { Knex } from 'knex';

// https://knexjs.org/guide/migrations.html#custom-migration-sources
export default class FsMigrationSource implements Knex.MigrationSource<string> {
  // eslint-disable-next-line node/no-process-env
  private readonly ext = process.env.NODE_ENV === 'production' ? 'js' : 'ts';
  private readonly directory: string;

  constructor(directory: string) {
    this.directory = directory;
  }
  // Must return a Promise containing a list of migrations.
  // Migrations can be whatever you want,
  // they will be passed as arguments to getMigrationName
  // and getMigration
  //
  // In our case migrations are filenames without extension
  async getMigrations(): Promise<string[]> {
    // In this example we are just returning migration names
    const dirContent = await fs.readdir(this.directory);
    // assume that migrations have extensions (== no directories with '.' symbol)
    return dirContent
      .filter((f) => f.includes('.'))
      .map((f) => f.split('.')[0])
      .sort();
  }

  // When restoring production dump on development, migration names will have '.js' extension
  getMigrationName(migration: string): string {
    return migration + '.js';
  }

  async getMigration(migration: string): Promise<Knex.Migration> {
    const moduleName = path.resolve(
      process.cwd(),
      this.directory,
      `${migration}.${this.ext}`,
    );
    return import(moduleName);
  }
}
