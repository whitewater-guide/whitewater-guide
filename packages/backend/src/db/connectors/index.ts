export * from './createConnectors';
export * from './OffsetConnector';
export * from './RelayConnector';
export * from './types';
