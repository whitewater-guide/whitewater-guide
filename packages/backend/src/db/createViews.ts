import { readFileSync } from 'fs';
import type { Knex } from 'knex';

import { resolveRelative } from '../utils/index';

export async function createView(db: Knex, view: string, version = 1) {
  for (let i = version; i >= 1; i -= 1) {
    const file = resolveRelative(
      __dirname,
      '../migrations',
      `${i.toString(10).padStart(3, '0')}/${view}_view.sql`,
    );
    let sql: string;
    try {
      sql = readFileSync(file).toString();
    } catch (e) {
      continue;
    }
    // should be created sequentially
    // eslint-disable-next-line no-await-in-loop
    await db.schema.raw(sql);
    return;
  }
  throw new Error(
    `No sql files for view '${view}' of version ${version} or lower found`,
  );
}

// Views will be created in reverse order
export async function createViews(
  db: Knex,
  version: number,
  ...views: string[]
) {
  const inverted = [...views].reverse();
  for (const view of inverted) {
    await createView(db, view, version);
  }
}
