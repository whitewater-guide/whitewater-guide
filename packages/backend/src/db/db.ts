/* eslint-disable node/no-process-env */
import Knex from 'knex';

import knexConfig from './knexfile';

const env = process.env.NODE_ENV || 'development';
const config: Knex.Config = knexConfig[env];

export const knex = Knex(config);

let testTransaction: Knex.Transaction | null = null;

export function db(getKnexInstance = false): Knex<any, any> {
  if (env === 'test' && !getKnexInstance) {
    if (testTransaction) {
      return testTransaction;
    }
    throw new Error('No saved test transaction found');
  }
  return knex;
}

// See here https://github.com/tgriesser/knex/issues/2076
// And here https://gist.github.com/odigity/7f37077de74964051d45c4ca80ec3250
export const holdTransaction = () => {
  if (env !== 'test') {
    throw new Error('holdTransaction is only available in test mode');
  }
  const tmp: any = {};
  const p = new Promise((resolve) => {
    tmp.resolve = resolve;
  });
  knex
    .transaction((tx) => {
      testTransaction = tx;
      tmp.resolve();
    })
    .catch(() => {
      // ignore
    });
  return p;
};

export const rollbackTransaction = async () => {
  if (env !== 'test') {
    throw new Error('rollbackTransaction is only available in test mode');
  }
  if (testTransaction) {
    await testTransaction.rollback();
    testTransaction = null;
  }
};
