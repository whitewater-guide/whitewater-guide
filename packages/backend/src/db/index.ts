import db from './db';

export default db;
export * from './updatedAtTrigger';
export * from './runSqlFile';
export * from './createViews';
export * from './dropViews';
export * from './rawUpsert';
export * from './types';
export * from './db';
export * from './queryBuilders';
