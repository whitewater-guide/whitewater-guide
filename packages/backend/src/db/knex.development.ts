/* eslint-disable node/no-process-env */
import type { Knex } from 'knex';

import FsMigrationSource from './FsMigrationSource';

const development = (): Knex.Config => ({
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST,
    port: 5432,
    database: process.env.POSTGRES_DB,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD,
    keepAlive: true,
  },
  migrations: {
    migrationSource: new FsMigrationSource('./src/migrations'),
    tableName: 'migrations',
  },
  seeds: {
    loadExtensions: ['.ts'],
    directory: './src/seeds/development',
  },
  debug: false,
});

export default development;
