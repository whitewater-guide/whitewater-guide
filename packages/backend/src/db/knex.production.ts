/* eslint-disable node/no-process-env */
import type { Knex } from 'knex';

import FsMigrationSource from './FsMigrationSource';

const production = (): Knex.Config => ({
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST,
    port: 5432,
    database: process.env.POSTGRES_DB,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD,
    keepAlive: true,
  },
  migrations: {
    tableName: 'migrations',
    migrationSource: new FsMigrationSource('./dist/migrations'),
  },
  pool: {
    min: 0,
  },
  // debug: true,
});

export default production;
