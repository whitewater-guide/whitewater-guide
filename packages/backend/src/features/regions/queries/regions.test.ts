import { anonContext, fakeContext, noTimestamps } from '@test';
import gql from 'graphql-tag';
import countBy from 'lodash/countBy';

import { holdTransaction, rollbackTransaction } from '~/db';
import {
  ADMIN,
  BOOM_USER_1500,
  EDITOR_GA_EC,
  EDITOR_GE,
  EDITOR_NO_EC,
  TEST_USER,
  TEST_USER2,
} from '~/seeds/test/01_users';
import {
  NUM_REGIONS,
  REGION_ECUADOR,
  REGION_GALICIA,
  REGION_GEORGIA,
  REGION_NORWAY,
} from '~/seeds/test/04_regions';

import {
  testCountGauges,
  testCountRegions,
  testCountSections,
  testListPremiumRegions,
  testListRegions,
  testListRegionsByName,
} from './regions.test.generated';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const _query = gql`
  query listRegions($page: Page) {
    regions(page: $page) {
      nodes {
        id
        name
        description
        season
        seasonNumeric
        bounds
        hidden
        premium
        sku
        editable
        license {
          slug
          name
          url
        }
        copyright
        createdAt
        updatedAt
        pois {
          id
          name
          description
          kind
          coordinates
        }
      }
      count
    }
  }
`;

describe('permissions', () => {
  it('editor should see his editable regions', async () => {
    // Cannot see hidden region Laos which he cannot edit
    // But can see hidden region Norway which he can edit
    const result = await testListRegions(undefined, fakeContext(EDITOR_NO_EC));
    expect(result.errors).toBeUndefined();
    expect(countBy(result.data?.regions.nodes, 'editable')).toMatchObject({
      true: 2,
      false: 4,
    });
  });

  it('editor should not see hidden regions without permission', async () => {
    const result = await testListRegions(undefined, fakeContext(EDITOR_GA_EC));
    expect(result.errors).toBeUndefined();
    expect(result.data?.regions.count).toBe(5);
    expect(result.data?.regions.nodes).not.toContainEqual(
      expect.objectContaining({ id: REGION_NORWAY }),
    );
  });

  it('anons should not see hidden regions', async () => {
    const result = await testListRegions({}, anonContext());
    expect(result.errors).toBeUndefined();
    expect(result.data?.regions.count).toBe(5);
    expect(result.data?.regions.nodes).toHaveLength(5);
    expect(result.data?.regions.nodes[0].hidden).toBe(null);
  });

  it('users should not see hidden regions', async () => {
    const result = await testListRegions({}, fakeContext(TEST_USER));
    expect(result.errors).toBeUndefined();
    expect(result.data?.regions.count).toBe(5);
    expect(result.data?.regions.nodes).toHaveLength(5);
    expect(result.data?.regions.nodes[0].hidden).toBe(null);
  });
});

describe('results', () => {
  it('admin should get all regions', async () => {
    const result = await testListRegions(undefined, fakeContext(ADMIN));
    expect(result.errors).toBeUndefined();
    expect(result.data?.regions.count).toBe(NUM_REGIONS);
    expect(result.data?.regions.nodes.length).toBe(NUM_REGIONS);
    expect(result.data?.regions.nodes.map(noTimestamps)).toMatchSnapshot();
  });

  it('should return rivers count', async () => {
    const _q = gql`
      query countRegions($page: Page) {
        regions(page: $page) {
          nodes {
            id
            name
            rivers {
              count
            }
          }
          count
        }
      }
    `;
    const result = await testCountRegions({}, fakeContext(ADMIN));
    expect(result.data?.regions.count).toBe(NUM_REGIONS);
    // Check name
    expect(result.data?.regions.nodes).toMatchObject([
      { rivers: { count: 1 } }, // Ecuador
      { rivers: { count: 2 } }, // galicia
      { rivers: { count: 1 } }, // georgia
      { rivers: { count: 0 } }, // laos
      { rivers: { count: 2 } }, // norway
      { rivers: { count: 0 } }, // other
      { rivers: { count: 1 } }, // russia
    ]);
  });

  it('should return gauges count', async () => {
    const _q = gql`
      query countGauges($page: Page) {
        regions(page: $page) {
          nodes {
            id
            name
            gauges {
              count
            }
          }
          count
        }
      }
    `;
    const result = await testCountGauges({}, fakeContext(ADMIN));
    expect(result.data?.regions).toBeDefined();
    expect(result.data?.regions.nodes).toMatchObject([
      { gauges: { count: 0 } }, // Ecuador
      { gauges: { count: 1 } }, // galicia
      { gauges: { count: 2 } }, // georgia
      { gauges: { count: 0 } }, // laos
      { gauges: { count: 1 } }, // norway
      { gauges: { count: 0 } }, // others
      { gauges: { count: 0 } }, // russia
    ]);
  });

  it('should return sections count', async () => {
    const _q = gql`
      query countSections($page: Page) {
        regions(page: $page) {
          nodes {
            id
            name
            sections {
              count
            }
          }
          count
        }
      }
    `;
    const result = await testCountSections({}, fakeContext(ADMIN));
    expect(result.errors).toBeUndefined();
    expect(result.data?.regions).toBeDefined();
    const regions = result.data?.regions.nodes;
    // At least one region should have some sections
    expect(regions?.some((r: any) => r.sections.count > 0)).toBe(true);
  });

  it('should search by name', async () => {
    const _q = gql`
      query listRegionsByName($filter: RegionFilterOptions, $page: Page) {
        regions(filter: $filter, page: $page) {
          nodes {
            id
            name
          }
        }
      }
    `;

    const result = await testListRegionsByName(
      { filter: { searchString: 'Or' } },
      fakeContext(ADMIN),
    );
    expect(result.errors).toBeUndefined();
    expect(result.data?.regions.nodes).toHaveLength(3);
    expect(result.data?.regions.nodes).toContainEqual(
      expect.objectContaining({
        id: REGION_NORWAY,
      }),
    );
    expect(result.data?.regions.nodes).not.toContainEqual(
      expect.objectContaining({
        id: REGION_GALICIA,
      }),
    );
  });
});

describe('i18n', () => {
  it('should be able to specify language', async () => {
    const result = await testListRegions({}, fakeContext(ADMIN, 'ru'));
    expect(result.data?.regions.count).toBe(NUM_REGIONS);
    expect(result.data?.regions.nodes).toContainEqual(
      expect.objectContaining({ name: 'Галисия' }),
    );
  });

  it('should fall back to english when not translated', async () => {
    const result = await testListRegions({}, fakeContext(ADMIN, 'ru'));
    expect(result.data?.regions.count).toBe(NUM_REGIONS);
    expect(result.data?.regions.nodes).toContainEqual(
      expect.objectContaining({ name: 'Norway' }),
    );
  });
});

describe('premium access', () => {
  const _q = gql`
    query listPremiumRegions($page: Page) {
      regions(page: $page) {
        nodes {
          id
          hasPremiumAccess
        }
        count
      }
    }
  `;

  it('without purchases', async () => {
    const result = await testListPremiumRegions({}, fakeContext(TEST_USER2));
    expect(result.data?.regions.nodes).not.toContainEqual(
      expect.objectContaining({ hasPremiumAccess: true }),
    );
  });

  it('with single region purchased', async () => {
    const result = await testListPremiumRegions(
      {},
      fakeContext(TEST_USER, 'en'),
    );
    expect(result.data?.regions.nodes).toEqual(
      expect.arrayContaining([
        { id: REGION_ECUADOR, hasPremiumAccess: false }, // invalid transaction
        { id: REGION_GEORGIA, hasPremiumAccess: true }, // valid transaction
      ]),
    );
  });

  it('with group of regions purchased', async () => {
    const result = await testListPremiumRegions(
      {},
      fakeContext(BOOM_USER_1500, 'en'),
    );
    expect(result.data?.regions.nodes).toEqual(
      expect.arrayContaining([
        { id: REGION_ECUADOR, hasPremiumAccess: false },
        { id: REGION_GEORGIA, hasPremiumAccess: true },
      ]),
    );
  });

  it('admin should always have premium access', async () => {
    const result = await testListPremiumRegions({}, fakeContext(ADMIN));
    expect(result.data?.regions.nodes).not.toContainEqual(
      expect.objectContaining({ hasPremiumAccess: false }),
    );
  });

  it('editor should always have premium access', async () => {
    const result = await testListPremiumRegions({}, fakeContext(EDITOR_GE));
    expect(result.data?.regions.nodes).toEqual(
      expect.arrayContaining([
        { id: REGION_ECUADOR, hasPremiumAccess: false },
        { id: REGION_GEORGIA, hasPremiumAccess: true },
      ]),
    );
  });
});
