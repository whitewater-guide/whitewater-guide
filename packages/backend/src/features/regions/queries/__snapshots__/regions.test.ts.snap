// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`results admin should get all regions 1`] = `
[
  {
    "bounds": [
      [
        1,
        0,
        0,
      ],
      [
        0,
        1,
        0,
      ],
      [
        1,
        1,
        0,
      ],
    ],
    "copyright": "copyrighter",
    "description": null,
    "editable": true,
    "hidden": false,
    "id": "2caf75ca-7625-11e7-b5a5-be2e44b06b34",
    "license": {
      "name": "Attribution-ShareAlike 4.0 International (CC BY-SA 4.0)",
      "slug": "CC_BY-SA",
      "url": "https://creativecommons.org/licenses/by-sa/4.0/",
    },
    "name": "Ecuador",
    "pois": [],
    "premium": true,
    "season": null,
    "seasonNumeric": [],
    "sku": "region.ecuador",
  },
  {
    "bounds": [
      [
        -114,
        46,
        0,
      ],
      [
        -115,
        46,
        0,
      ],
      [
        -115,
        47,
        0,
      ],
      [
        -114,
        47,
        0,
      ],
    ],
    "copyright": null,
    "description": "Description of Galicia",
    "editable": true,
    "hidden": false,
    "id": "bd3e10b6-7624-11e7-b5a5-be2e44b06b34",
    "license": null,
    "name": "Galicia",
    "pois": [
      {
        "coordinates": [
          20,
          30,
          40,
        ],
        "description": "Galicia Point 1 description",
        "id": "573f995a-d55f-4faf-8f11-5a6016ab562f",
        "kind": "kayak-shop",
        "name": "Galicia Point 1",
      },
      {
        "coordinates": [
          11,
          22,
          33,
        ],
        "description": "Galicia Point 2 description",
        "id": "d7530317-efac-44a7-92ff-8d045b2ac893",
        "kind": "other",
        "name": "Galicia Point 2",
      },
    ],
    "premium": false,
    "season": "Late autumn - early spring",
    "seasonNumeric": [
      20,
      21,
    ],
    "sku": null,
  },
  {
    "bounds": [
      [
        43.51,
        39.97,
        0,
      ],
      [
        41.93,
        46.39,
        0,
      ],
      [
        41.05,
        46.52,
        0,
      ],
      [
        41.51,
        41.58,
        0,
      ],
    ],
    "copyright": null,
    "description": "description of Georgia",
    "editable": true,
    "hidden": false,
    "id": "8e119768-37f3-11e8-b467-0ed5f89f718b",
    "license": null,
    "name": "Georgia",
    "pois": [],
    "premium": true,
    "season": "spring, summer, autumn",
    "seasonNumeric": [
      11,
      12,
    ],
    "sku": "region.georgia",
  },
  {
    "bounds": [
      [
        20.72,
        80.38,
        0,
      ],
      [
        14.2,
        74.58,
        0,
      ],
      [
        15.05,
        71.9,
        0,
      ],
      [
        22.66,
        77.5,
        0,
      ],
    ],
    "copyright": null,
    "description": "laos description",
    "editable": true,
    "hidden": true,
    "id": "a84d7eda-37f3-11e8-b467-0ed5f89f718b",
    "license": null,
    "name": "Laos",
    "pois": [
      {
        "coordinates": [
          4,
          3,
          343,
        ],
        "description": "Laos point 1 description",
        "id": "a52dfcc6-3716-11e8-b467-0ed5f89f718b",
        "kind": "other",
        "name": "Laos Point 1",
      },
    ],
    "premium": false,
    "season": "laos season",
    "seasonNumeric": [
      3,
      4,
      5,
      6,
    ],
    "sku": null,
  },
  {
    "bounds": [
      [
        4,
        5,
        0,
      ],
      [
        7,
        8,
        0,
      ],
      [
        5,
        6,
        0,
      ],
    ],
    "copyright": null,
    "description": null,
    "editable": true,
    "hidden": true,
    "id": "b968e2b2-76c5-11e7-b5a5-be2e44b06b34",
    "license": {
      "name": "Proprietary license",
      "slug": null,
      "url": null,
    },
    "name": "Norway",
    "pois": [],
    "premium": true,
    "season": null,
    "seasonNumeric": [],
    "sku": "region.norway",
  },
  {
    "bounds": [
      [
        -179,
        -89,
        0,
      ],
      [
        -179,
        89,
        0,
      ],
      [
        179,
        89,
        0,
      ],
      [
        179,
        -89,
        0,
      ],
    ],
    "copyright": null,
    "description": null,
    "editable": true,
    "hidden": false,
    "id": "7e535c02-8624-46a1-a865-d88fe4557361",
    "license": null,
    "name": "Others",
    "pois": [],
    "premium": false,
    "season": null,
    "seasonNumeric": [],
    "sku": null,
  },
  {
    "bounds": [
      [
        -17,
        -89,
        0,
      ],
      [
        -17,
        89,
        0,
      ],
      [
        17,
        89,
        0,
      ],
      [
        17,
        -89,
        0,
      ],
    ],
    "copyright": null,
    "description": null,
    "editable": true,
    "hidden": false,
    "id": "a2056850-23ae-434d-b06c-7ea1900e5de2",
    "license": null,
    "name": "Россия",
    "pois": [],
    "premium": false,
    "season": null,
    "seasonNumeric": [],
    "sku": null,
  },
]
`;
